lib/puppet/pops/evaluator/closure.rb

```diff
@@ -48,18 +48,39 @@ def call(*args)
   # checks of the argument type/arity validity).
   # @api private
   def invoke(instance, calling_scope, args, &block)
-    call(*args, &block)
+    scopes_to_restore = @enclosing_scope.unset_ephemeral_var(1)
+    begin
+      # Only global scope should be visible to the called function. Other ephemeral scopes must be popped
+      # and later restored
+      call(*args, &block)
+    ensure
+      unless scopes_to_restore.nil?
+        @enclosing_scope.unset_ephemeral_var(1)
```

Comment:
The call_by_name does what I expect, it takes the level and restores it. Thats it.
Not sure what the restore stuff is doing. Remember that the match scopes are opaque - if you try to lookup say $9 and the closest matchscope has an empty match then the search stops there. Thus, it is just a matter of assigning a match to the current matchscope, or pushing one (empty or with a match), then, at the end pop back to the level that was remembered on entry. 

But maybe I am reading the code wrong and I am being confused of what invoke does.